src/transformers/optimization.py

```diff
@@ -940,6 +943,325 @@ def get_lr(self):
+    """
+    Implements Adam Mini algorithm as introduced in [Adam-mini: Use Fewer Learning Rates To Gain More]
+    (https://arxiv.org/abs/2406.16793).
+
+    Parameters:
+        model (`Union[PreTrainedModel, nn.Module]`):
+            The model being trained.
+        lr (`float`, *optional*, defaults to 0.001):
+            The learning rate to use.
+        betas (`Tuple[float,float]`, *optional*, defaults to `(0.9, 0.999)`):
+            Adam's betas parameters (b1, b2).
+        eps (`float`, *optional*, defaults to 1e-06):
+            Adam's epsilon for numerical stability.
+        weight_decay (`float`, *optional*, defaults to 0.0):
+            Decoupled weight decay to apply.
+        model_sharding (`bool`, *optional*, defaults to `False`):
+            Set to True if you are using model parallelism with more than 1 GPU, including FSDP and
+            zero_1,2,3 in Deepspeed. Set to False if otherwise.
+        n_feature (`int`, *optional*, defaults to `None`):
+            Dimension for hidden features. Can be left unspecified if training non-transformer models.
+        n_head (`int`, *optional*, defaults to `None`):
+            Number of attention heads. Can be left unspecified if training non-transformer models.
+        n_kv_head (`int`, *optional*, defaults to `None`):
+            Number of heads for Key and Value. Or equivalently, number of query groups in Group Query
+            Attention. Also known as "n_query_groups".  If not specified, it will be equal to n_head.
+            Can be left unspecified if training non-transformer models.
+    """
+
+    def __init__(
+        self,
+        model: Union[PreTrainedModel, nn.Module],
+        lr: float = 1e-3,
+        betas: Tuple[float, float] = (0.9, 0.999),
+        eps: float = 1e-6,
+        weight_decay: float = 0.0,
+        model_sharding: bool = False,
+        n_feature: int = None,
+        n_head: int = 32,
+        n_kv_head: int = None,
+    ):
+        if lr < 0.0:
+            raise ValueError(f"Invalid learning rate: {lr} - should be >= 0.0")
+        if not 0.0 <= betas[0] < 1.0:
+            raise ValueError(f"Invalid beta parameter: {betas[0]} - should be in [0.0, 1.0)")
+        if not 0.0 <= betas[1] < 1.0:
+            raise ValueError(f"Invalid beta parameter: {betas[1]} - should be in [0.0, 1.0)")
+        if not 0.0 <= weight_decay:
+            raise ValueError(f"Invalid weight_decay value: {weight_decay} - should be >= 0.0")
+        if not 0.0 <= eps:
+            raise ValueError(f"Invalid epsilon value: {eps} - should be >= 0.0")
+
+        self.n_feature = n_feature
+        self.n_head = n_head
+        if n_kv_head is not None:
+            self.n_kv_head = n_kv_head
+            if self.n_head % self.n_kv_head != 0:
+                raise ValueError("n_head is not a multiple of n_kv_head")
+        else:
+            self.n_kv_head = self.n_head
+        self.model = model
+        self.world_size = torch.cuda.device_count()
```

Comment:
Since all distributed stuff runs via `accelerate` to be device-agnostic, we should check here with it instead. 
```suggestion
        self.world_size = PartialState().num_processes
```

Reply:
(Note the import for earlier is:
```python
if is_accelerate_available():
    from accelerate import PartialState
```